sys/linux/dev_qat_adf_ctl.txt

```diff
@@ -0,0 +1,108 @@
+# Copyright 2020 syzkaller project authors. All rights reserved.
+# Use of this source code is governed by Apache 2 LICENSE that can be found in the LICENSE file.
+
+# NEED: These descriptions need an actual Intel QAT device or an emulation so we can fuzz deeper into the subsystem.
+# /dev/qat_adf_ctl is only the control interface, and currently ioctl$IOCTL_GET_NUM_DEVICES returns 0 on qemu.
+
+include <linux/fcntl.h>
+include <drivers/crypto/qat/qat_common/adf_cfg_user.h>
+include <drivers/crypto/qat/qat_common/adf_cfg_common.h>
+resource fd_qat[fd]
+
+openat$qat_adf_ctl(fd const[AT_FDCWD], file ptr[in, string["/dev/qat_adf_ctl"]], flags flags[open_flags], mode const[0]) fd_qat
+ioctl$IOCTL_CONFIG_SYS_RESOURCE_PARAMETERS(fd fd_qat, cmd const[IOCTL_CONFIG_SYS_RESOURCE_PARAMETERS], arg ptr[in, adf_user_cfg_ctl_data])
+ioctl$IOCTL_STOP_ACCEL_DEV(fd fd_qat, cmd const[IOCTL_STOP_ACCEL_DEV], arg ptr[in, adf_user_cfg_ctl_data])
+ioctl$IOCTL_START_ACCEL_DEV(fd fd_qat, cmd const[IOCTL_START_ACCEL_DEV], arg ptr[in, adf_user_cfg_ctl_data])
+ioctl$IOCTL_GET_NUM_DEVICES(fd fd_qat, cmd const[IOCTL_GET_NUM_DEVICES], arg ptr[out, int32])
+ioctl$IOCTL_STATUS_ACCEL_DEV(fd fd_qat, cmd const[IOCTL_STATUS_ACCEL_DEV], arg ptr[inout, adf_dev_status_info])
+
+adf_user_cfg_ctl_data {
+	config_section	adf_ptr64[adf_user_cfg_section]
+	device_id	int8
+} [packed]
+
+# Section
+
+adf_user_cfg_section {
+	name	string[qat_section_name, ADF_CFG_MAX_SECTION_LEN_IN_BYTES]
+	params	adf_ptr64[adf_user_cfg_key_val]
+	next	adf_ptr64[adf_user_cfg_section]
+} [packed]
+
+qat_section_name = "GENERAL", "KERNEL", "Accelerator", "Accelerator0", "Accelerator1"
+
+# Key-values
+
+adf_user_cfg_key_val [
+	adf_dec	adf_user_cfg_key_val_t[ADF_DEC, adf_user_cfg_val_dec]
+	adf_hex	adf_user_cfg_key_val_t[ADF_HEX, adf_user_cfg_val_hex]
+	adf_str	adf_user_cfg_key_val_t[ADF_STR, adf_user_cfg_val_str]
+]
+
+type adf_user_cfg_key_val_t[TYPE, VAL] {
+	key	adf_user_cfg_key
+	val	VAL
+	next	adf_ptr64[adf_user_cfg_key_val]
```

Comment:
The syntax is correct, however it is better if you include` next` and` type` attributes in `adf_user_cfg_key_val `structure while you define `adf_dec,` `adf_hex`, `adf_str` as unions with `key and val `again

Reply:
Could you explain more on what you meant? Do you mean changing the type template so that each member has its corresponding type argument? 
```
type adf_user_cfg_key_val_t[KEY, TYPE, VAL, NEXT] {
	key	KEY
	val	VAL
	next	NEXT
	type	const[TYPE, int32]
} [packed]
```
and define `adf_user_cfg_key_val` like
```
adf_user_cfg_key_val [
	adf_dec	adf_user_cfg_key_val_t[adf_user_cfg_key, ADF_DEC, adf_user_cfg_val_dec, adf_ptr64[adf_user_cfg_key_val]]
	adf_hex	adf_user_cfg_key_val_t[adf_user_cfg_key, ADF_HEX, adf_user_cfg_val_hex, adf_ptr64[adf_user_cfg_key_val]]
...
]
```
I am not sure if this will increase readability because it seems like a lot of item is repeated.

I searched through the existing descriptions and saw some that is similar to what I wrote. For example in vnet.txt:
```
ipv6_packet [
        generic         ipv6_packet_t[flags[ipv6_types, int8], array[int8]]
        tcp             ipv6_packet_t[const[IPPROTO_TCP, int8], tcp_packet]
...
] [varlen]

type ipv6_packet_t[PROTO, PAYLOAD] {
        priority        int8:4
        version         const[6, int8:4]
        flow_label      array[int8, 3]
        length          len[payload, int16be]
        next_header     PROTO
        hop_limit       flags[hop_limits, int8]
        src_ip          ipv6_addr
        dst_ip          ipv6_addr
        payload         ipv6_packet_payload[PAYLOAD]
} [packed]
```

Reply:
I was meaning the following. But I've realised that this needs a change in ordering which is wrong. Yours is just fine; you can resolve the comment.

```
adf_user_cfg_key_val{
        key     adf_user_cfg_key
        next	adf_ptr64[adf_user_cfg_key_val]
        gen     adf_user_cfg_key_val_gen
}
adf_user_cfg_key_val_gen [
	adf_dec	adf_user_cfg_key_val_t[ADF_DEC, adf_user_cfg_val_dec]
	adf_hex	adf_user_cfg_key_val_t[ADF_HEX, adf_user_cfg_val_hex]
	adf_str	adf_user_cfg_key_val_t[ADF_STR, adf_user_cfg_val_str]
]

type adf_user_cfg_key_val_t[TYPE, VAL] {
	val	VAL
	type	const[TYPE, int32] 
} [packed]

```